crypto/fipsmodule/rand/new_rand.c

```diff
@@ -0,0 +1,290 @@
+
+  OPENSSL_cleanse(seed, CTR_DRBG_ENTROPY_LEN);
+  OPENSSL_cleanse(personalization_string, CTR_DRBG_ENTROPY_LEN);
+}
+
+// RAND_bytes_core generates |out_len| bytes of randomness and puts them in
+// |out|. The CTR-DRBG state |state| is managed to ensure uniqueness and usage
+// requirements are met.
+//
+// The argument |use_user_pred_resistance| must be either
+// |RAND_USE_USER_PRED_RESISTANCE| or |RAND_NO_USER_PRED_RESISTANCE|. The former
+// cause the content of |user_pred_resistance| to be mixed in as prediction
+// resistance. The latter ensures that |user_pred_resistance| is not used.
+static void RAND_bytes_core(
+  struct rand_thread_local_state *state,
+  uint8_t *out, size_t out_len,
+  const uint8_t user_pred_resistance[RAND_PRED_RESISTANCE_LEN],
+  int use_user_pred_resistance) {
+
+  // Ensure CTR-DRBG state is unique.
+  if (rand_ensure_ctr_drbg_uniquness(state, out_len) != 1) {
+    rand_ctr_drbg_reseed(state);
+  }
+
+  // If a prediction resistance source is available, use it.
+  // Prediction resistance is only used on first invocation of the CTR-DRBG,
+  // ensuring that its state is randomized before using output.
+  size_t first_pred_resistance_len = 0;
+  uint8_t pred_resistance[RAND_PRED_RESISTANCE_LEN] = {0};
+  rand_maybe_get_ctr_drbg_pred_resistance(state, pred_resistance,
+    &first_pred_resistance_len);
+
+  // If caller input user-controlled prediction resistance, use it.
+  if (use_user_pred_resistance == RAND_USE_USER_PRED_RESISTANCE) {
+    for (size_t i = 0; i < RAND_PRED_RESISTANCE_LEN; i++) {
+      pred_resistance[i] ^= user_pred_resistance[i];
+    }
+    first_pred_resistance_len = RAND_PRED_RESISTANCE_LEN;
+  }
+
+  assert(first_pred_resistance_len == 0 ||
+         first_pred_resistance_len == RAND_PRED_RESISTANCE_LEN);
+
+  // Iterate CTR-DRBG generate until we generated |out_len| bytes of randomness.
+  while (out_len > 0) {
+    size_t todo = out_len;
+    if (todo > CTR_DRBG_MAX_GENERATE_LENGTH) {
+      todo = CTR_DRBG_MAX_GENERATE_LENGTH;
+    }
+
+    // Each reseed interval can generate up to
+    // |CTR_DRBG_MAX_GENERATE_LENGTH*2^{kCtrDrbgReseedInterval}| bytes.
+    // Determining the time(s) to reseed prior to entering the CTR-DRBG generate
+    // loop is a doable strategy. But tracking reseed times adds unnecessary
+    // complexity. Instead our strategy is optimizing for simplicity.
+    // |out_len < CTR_DRBG_MAX_GENERATE_LENGTH| will be the majority case
+    // (by far) and requires a single check in either strategy.
+    // Note if we just reseeded through |rand_ctr_drbg_reseed()|, we won't
+    // reseed again here.
+    if( state->generate_calls_since_seed + 1 >= kCtrDrbgReseedInterval) {
+      rand_ctr_drbg_reseed(state);
+    }
+
+    if (!CTR_DRBG_generate(&state->drbg, out, todo, pred_resistance,
+          first_pred_resistance_len)) {
+      abort();
+    }
+
+    out += todo;
+    out_len -= todo;
+    state->generate_calls_since_seed++;
+    first_pred_resistance_len = 0;
```

Comment:
Ohhhh, I think I see what your comment is talking about: if we need to call the ctr drbg multiple times for a single call to RAND_bytes we only use prediction resistance on the first call.